File: src_Lua/in_out.lua
t = {}

-- Faz a leitura do arquivo que contém a distância limite.
function t.readFileOfDistance(fileName)
    file = io.open(fileName,"r")
    distance = file:read("*n")
    file:close(file)
    return distance
end

-- Faz a leitura do arquivo das coordenadas dos pontos
-- e retorna o conjunto de pontos.
function t.readFileOfPointsCoord(fileName)
    X = {} -- Conjunto de pontos.
    i = 1  -- Índice da linha do arquivo que representa o ponto.
    for line in io.lines(fileName) do
        p = {}
        p.coord = {}
        p.line = i
        for n in string.gmatch(line,"[^ ]+") do
            table.insert(p.coord, tonumber(n))
        end
        table.insert(X,p)
        i = i+1
    end
    return X
end

-- Cria um arquivo e escreve a SSE do agrupamento.
function t.writeResultFile(fileName, sse)
    file = io.open(fileName,"w")
    file:write(string.format("%.4f", sse))
    file:close(file)
end

-- Cria um arquivo e escreve os grupos formados no agrupamento.
function t.writeGroupsFile(fileName, C)
    file = io.open(fileName,"w")
    
    for i,group in pairs(C) do
        for k,v in pairs(group) do
            file:write(v.line)
            if k ~= (#group) then
                file:write(" ")
            end
        end
        if i ~= #C then
            file:write("\n\n")
        end
    end
end

return t

File: src_Rust/in_out.rs
use std::io::{BufRead, BufReader, BufWriter, Write};
use std::fs::File;
use std::fs;

use point::{Point, Group, display_groups};

// Faz a leitura do arquivo que contém a distância limite.
pub fn read_distance(f: String) -> f64 {
    let reader = fs::read_to_string(f).expect("Erro na abertura do arquivo da distancia.");
    return reader.trim().parse::<f64>().unwrap();
}

// Faz a leitura do arquivo das coordenadas dos pontos e retorna o vetor de pontos.
pub fn read_points_coord(f: String) -> Vec<Point> {
    let reader = BufReader::new(File::open(f).expect("Erro na abertura do arquivo das coordenadas de pontos."));

    let mut points: Vec<Point> = Vec::new();

    for (i,cds) in reader.lines().enumerate() {
        let mut coords: Vec<f64> = Vec::new();
        for c in cds.unwrap().split_whitespace() {
            coords.push(c.parse::<f64>().unwrap());
        }
        points.push(Point{coord: coords, line: (i+1) as i32});
    } 
    return points;
}

// Cria um arquivo e escreve a SSE do agrupamento.
pub fn write_result(f: String, sse: f64) {
    let mut writer = BufWriter::new(File::create(f).expect("Erro ao criar o arquivo result.txt"));
    write!(writer, "{:.4}", sse).unwrap();
}

// Cria um arquivo e escreve os grupos formados no agrupamento.
pub fn write_groups(f: String, groups: Vec<Group>) {
    let mut writer = BufWriter::new(File::create(f).expect("Erro ao criar o arquivo saida.txt"));
    write!(writer, "{}", display_groups(&groups)).unwrap();
}